staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,6 +285,247 @@ func TestSimpleQueue(t *testing.T) {
+// only when the item isn't already queued.  This fuctionality also applies to immediate adds made by the AddWithOptions
+// method.
+func TestOptionsTakeLongerTakeExisting(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+	third := "bar"
+
+	// add an item into the waiting queue.
+	q.AddAfter(first, 2*time.Second)
+
+	// attempt to immediately add an item using TakeLonger option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeLonger})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to immediately add an item using TakeExisting option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to add an item not queued should queue successfully...
+	q.AddWithOptions(third, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{third},
+		waiting: []interface{}{first},
+	})
+
+	// step past the original delayed item and process it from the active queue
+	fakeClock.Step(5 * time.Second)
+	if err := waitForAdded(q, 2); err != nil {
+		t.Fatalf("the first item should have been queued")
+	}
+	q.AssertState(t, QueueState{
+		active: []interface{}{first, third},
+	})
+	// pop them both off the queue
+	item, _ := q.Get()
+	q.Done(item)
+	item, _ = q.Get()
+	q.Done(item)
+
+	// add the second item again with DropIfWaiting which should now succeed
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		active: []interface{}{second},
+	})
+}
+
+// TestDelayedAddInsertSchemes - tests the differnent scheduling behaviours available when the same item is already waiting.
+func TestDelayedAddInsertSchemes(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+
+	// TakeShorter
+	// add an item into the waiting queue.
+	q.AddAfter(first, 5*time.Second)
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddWithOptions(second, DelayingOptions{Duration: 1 * time.Second, WhenWaiting: TakeShorter})
+	// step past the shorter item and confirm that it got queued.
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+	item, _ := q.Get()
+	q.Done(item)
+
+	// TakeLonger
+	// add an item into the waiting queue.
+	q.AddAfter(first, 5*time.Second)
+	// attempt add delayed again with a shorter time time out and TakeLonger option.
+	q.AddWithOptions(second, DelayingOptions{Duration: 1 * time.Second, WhenWaiting: TakeLonger})
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
```

Comment:
this test takes too long with these timeouts. Why not just `AssertState` here and then forward the clock again?

Reply:
A wait is needed to allow the "waitingLoop" to fire and process the item before we check the desired state.  This is in every test when stepping the clock forward and expecting something to happen.  In this case, the wait is usually very short as this tests for a positive result.  There are a couple of tests which must wait for the timeout in order to confirm something didn't happen.  This isn't perfect but its how things are done in this test suit and I don't want to make any more changes to it.